modules/python/src2/hdr_parser.py

```diff
@@ -911,9 +911,12 @@ def parse(self, hname, wmode=True):
                         if stmt_type.startswith("enum"):
                             decls.append([stmt_type + " " + self.get_dotted_name(name), "", [], decl, None, ""])
                         else:
-                            decls.append(decl)
+                            if "cv.cudacodec.VideoReader.nextFrame" not in decl[0]:
+                                # Disable Mat decl for specific cudacodec function which can take zero args, to allow python call to
+                                # fall through to correct GpuMat decl
```

Comment:
Try this instead:
```
ret, gpu_mat = .nextFrame(frame=cv.cuda_GpuMat())
```

or add separate wrapper for GPU-only (remove WRAP from `nextFrame(OutputArray)`):
```
CV_WRAP inline bool nextFrame(CV_OUT GpuMat& frame) { return nextFrame(OutputArray(frame); }
```

Reply:
I have tried

> CV_WRAP inline bool nextFrame(CV_OUT GpuMat& frame, Stream& stream = Stream::Null()) { return nextFrame(OutputArray(frame), stream); }

because it needs to work with streams.  The problem is I can't figure out how to generate bindings so both 

> ret,gpu_mat = _reader.nextFrame()

and

> ret,_ = _reader.nextFrame(gpu_mat)

work.

The first case really needs to work because it is really useful to be able to interrogate the size of the expected output `GPUMat` when using the python interface.  If not then it can just silently not work because the size is wrong.

The second case definitely needs to work otherwise python will allocate a new `GPUMat` every time a video frame is read.

Reply:
@alalek 
The latest changes to the python bindings generation code are to ensure that

> CV_WRAP inline bool nextFrame(CV_OUT GpuMat& frame, Stream& stream = Stream::Null()) { return nextFrame(frame, stream); }

will work with both

> ret,_ = _reader.nextFrame(gpu_mat)

and

> ret,gpu_mat = _reader.nextFrame()

Reply:
Please run it locally - it doesn't work.

> test_cudacodec (test_cuda.cuda_test) ...
OpenCV(4.1.2-dev) Error: The function/feature is not implemented (getGpuMat is available only for cuda::GpuMat and cuda::HostMem) in getGpuMat, file modules/core/src/matrix_wrap.cpp, line 359
skipped 'NVCUVID is not installed'

(generated overloads are in this order: "Mat", "GpuMat", "UMat")

This works somehow:

```
ret,gpu_mat = _reader.nextFrame(cv.cuda_GpuMat())
self.assertTrue(ret)
print(type(gpu_mat))  # <== add check for this
#print(cv.utils.dumpInputArray(gpu_mat)) # - no support for GpuMat
```

Reply:
Is that with the change to 

> CV_WRAP inline bool nextFrame(CV_OUT GpuMat& frame, Stream& stream = Stream::Null()) { return nextFrame(frame, stream); }

Reply:
Got it.
Could you please create PR into opencv_contrib too (branch "fix_cudacodec_python")?

Reply:
@alalek 
Would it make more sense to have an additional flag to signal that the `OutputArray` is a `GpuMat` instead of adding additional overides?  For example adding `CV_GPU`

` CV_WRAP virtual bool nextFrame(CV_GPU OutputArray frame, Stream &stream = Stream::Null()) = 0;`

and excluding the Mat bindings when it is present

```
if "CV_GPU" in arg_str:
    modlist.append("/G")
    arg_str = arg_str.replace("CV_GPU", "")
```

```
has_gpu_mat = any(flag == '/G' for var in decl[3] for flag in var[3])
if not has_gpu_mat:
    decls.append(decl)
```

I guess this could them be extended to work if there was a mixture of array types to a function.

Reply:
> excluding the Mat bindings

Perhaps reordering should be better.

---

> `if "CV_GPU" in arg_str:`

Not sure if we need separate notation for that. Lets try to reuse this condition:
```
if self._generate_gpumat_decls and "cv.cuda" in decl[0]:
```

Reply:
Can you expand on what you mean by reordering?  If we move the `Mat` bindings below `GpuMat ` then `GpuMat.download()` will fail.

Reply:
> move the Mat bindings below GpuMat

Yes, I mean this. `Mat` overload below `GpuMat`

---

> then GpuMat.download() will fail

Is this for case with result of `Mat` type? (so we need to remove Mat completely)
This make sense for outputs (like `OutputArray`), but perhaps we should accept `InputArray` as numpy (`Mat`) array. Anyway, lets try with "excluding the Mat bindings".

Reply:
Yes the problem is that `OutputArray` could be a `Mat` or `GpuMat` and currently there is no way to tell until the input is parsed.  This works fine if the python function has input arguments because the correct bindings will be chosen when this happens.  

Currently this also works when there are no input arguments because `GpuMat.download()` is the only function this applies to, so returning a `Mat` works.  Unfortunately this fails for `nextFrame()`.  Therefore if we reorder it will work for `nextFrame()` but fail for `GpuMat.download()`.

> This make sense for outputs (like OutputArray), but perhaps we should accept InputArray as numpy (Mat) array. Anyway, lets try with "excluding the Mat bindings".

I agree and I think that signaling that `OutputArray` is a `GpuMat` will allow input arrays which are a mixture of both types to be processed.  Currently I don't think there are any cuda functions which take a mixture of `Mat` and `GpuMat` but I may be wrong.